### DISCUSSION TITLE
Downloading G-code and Real-time g-code position

### DISCUSSION THREAD

**Discussion by Moosketair** [greghesp/ha-bambulab#544] [Q&A] [answered]
Hello,

Is it possible to download the G-code used for a print job? Additionally, I am interested in knowing if there is a way to obtain the current position in the G-code during an ongoing print, like on OctoPrint with the G-code Viewer plugin. My goal is to create a real-time visualization to track the progress of the print job.

**Comment by AdrianGarside** [comment, marked as answer]
Downloading the gcode is possible. Realtime tracking of it is not. The only data we get in the mqtt payload for progress is the layer count.

**Reply by afzanaz3** [reply]
Just came across this. I have a Bambu A1, and I am trying to set up an automated tracker for print time, energy usage, and material usage. Since there is no sensor currently for tracking material, and I don't have an AMS, I wanted to understand how to pull the Gcode and try to get the material estimate from that. Or if there's a smarter way to achieve this

**Reply by rsocko** [reply]
How would I pull the gcode? I'm not seeing an entity. Is it via a command?

**Reply by greghesp** [reply]
There isn't an easy way to pull the gcode per say.  What we do is try to identify the gcode file on the printer (if it exists) based on the Mqtt info.  If it's not there, we then try and download the corresponding 3mf file from the printer FTP server, extract it and then get the gcode out of it that way
